pkg/apis/mysql/v1alpha1/types.go

```diff
@@ -0,0 +1,393 @@
+// Copyright 2018 Oracle and/or its affiliates. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// The default MySQL version to use if not specified explicitly by user
+	defaultVersion      = "8.0.11"
+	defaultReplicas     = 1
+	defaultBaseServerID = 1000
+	// Max safe value for BaseServerID calculated as max MySQL server_id value - max Replication Group size
+	maxBaseServerID uint32 = 4294967295 - 9
+)
+
+const (
+	// MaxInnoDBClusterMembers is the maximum number of members supported by InnoDB
+	// group replication.
+	MaxInnoDBClusterMembers = 9
+
+	// ClusterNameMaxLen is the maximum supported length of a
+	// Cluster name.
+	// See: https://bugs.mysql.com/bug.php?id=90601
+	ClusterNameMaxLen = 28
+)
+
+// TODO (owain) we need to remove this because it's not reasonable for us to maintain a list
+// of all the potential MySQL versions that can be used and in reality, it shouldn't matter
+// too much. The burden of this is not worth the benfit to a user
+var validVersions = []string{
+	defaultVersion,
+}
+
+// ClusterSpec defines the attributes a user can specify when creating a cluster
+type ClusterSpec struct {
+	// Version defines the MySQL Docker image version.
+	Version string `json:"version"`
+
+	// Replicas defines the number of running MySQL instances in a cluster
+	Replicas int32 `json:"replicas,omitempty"`
```

Comment:
Controversially (perhaps) I'd like to change this to `Members`. `Replicas` is semantically incorrect for the default and recommended deployment (i.e. single-primary mode). The MySQL documentation discusses cluster "members" and I think we should align with that.

//cc @owainlewis @garthy @jhorwit2